// @pushprotocol/restapi/src/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALPHA_FEATURE_CONFIG = exports.VIEM_CONFIG = exports.CHANNEL_TYPE = exports.ABIS = exports.MIN_TOKEN_BALANCE = exports.TOKEN_VIEM_NETWORK_MAP = exports.TOKEN = exports.CORE_CONFIG = exports.VIEM_CORE_CONFIG = exports.CHAIN_NAME = exports.CHAIN_ID = exports.ALIAS_CHAIN_ID = exports.ETH_CHAIN_ID = exports.API_BASE_URL = void 0;
const constants_1 = require("./constants");
const core_1 = require("./abis/core");
const comm_1 = require("./abis/comm");
const token_1 = require("./abis/token");
const chains_1 = require("viem/chains");
const customChains_1 = require("./customChains");
// for methods not needing the entire config
exports.API_BASE_URL = {
    [constants_1.ENV.PROD]: 'https://backend.epns.io/apis',
    [constants_1.ENV.STAGING]: 'https://backend-staging.epns.io/apis',
    [constants_1.ENV.DEV]: 'https://backend-dev.epns.io/apis',
    /**
     * **This is for local development only**
     */
    [constants_1.ENV.LOCAL]: 'http://localhost:4000/apis',
};
const BLOCKCHAIN_NETWORK = {
    ETH_MAINNET: 'eip155:1',
    ETH_SEPOLIA: 'eip155:11155111',
    POLYGON_MAINNET: 'eip155:137',
    POLYGON_AMOY: 'eip155:80002',
    BSC_MAINNET: 'eip155:56',
    BSC_TESTNET: 'eip155:97',
    OPTIMISM_TESTNET: 'eip155:11155420',
    OPTIMISM_MAINNET: 'eip155:10',
    POLYGON_ZK_EVM_TESTNET: 'eip155:2442',
    POLYGON_ZK_EVM_MAINNET: 'eip155:1101',
    ARBITRUM_TESTNET: 'eip155:421614',
    ARBITRUMONE_MAINNET: 'eip155:42161',
    FUSE_TESTNET: 'eip155:123',
    FUSE_MAINNET: 'eip155:122',
    BERACHAIN_TESTNET: 'eip155:80085',
    LINEA_MAINNET: 'eip155:59144',
    LINEA_TESTNET: 'eip155:59141',
    CYBER_CONNECT_TESTNET: 'eip155:111557560',
    CYBER_CONNECT_MAINNET: 'eip155:7560',
    BASE_TESTNET: 'eip155:84532',
    BASE_MAINNET: 'eip155:8453',
};
exports.ETH_CHAIN_ID = {
    [constants_1.ENV.PROD]: 1,
    [constants_1.ENV.STAGING]: 11155111,
    [constants_1.ENV.DEV]: 11155111,
    [constants_1.ENV.LOCAL]: 11155111,
};
exports.ALIAS_CHAIN_ID = {
    POLYGON: {
        [constants_1.ENV.PROD]: 137,
        [constants_1.ENV.STAGING]: 80002,
        [constants_1.ENV.DEV]: 80002,
        [constants_1.ENV.LOCAL]: 80002,
    },
    BSC: {
        [constants_1.ENV.PROD]: 56,
        [constants_1.ENV.STAGING]: 97,
        [constants_1.ENV.DEV]: 97,
        [constants_1.ENV.LOCAL]: 97,
    },
    OPTIMISM: {
        [constants_1.ENV.PROD]: 10,
        [constants_1.ENV.STAGING]: 11155420,
        [constants_1.ENV.DEV]: 11155420,
        [constants_1.ENV.LOCAL]: 11155420,
    },
    POLYGONZKEVM: {
        [constants_1.ENV.PROD]: 1101,
        [constants_1.ENV.STAGING]: 2442,
        [constants_1.ENV.DEV]: 2442,
        [constants_1.ENV.LOCAL]: 2442,
    },
    ARBITRUMONE: {
        [constants_1.ENV.PROD]: 42161,
        [constants_1.ENV.STAGING]: 421614,
        [constants_1.ENV.DEV]: 421614,
        [constants_1.ENV.LOCAL]: 421614,
    },
    FUSE: {
        [constants_1.ENV.PROD]: 122,
        [constants_1.ENV.STAGING]: 123,
        [constants_1.ENV.DEV]: 123,
        [constants_1.ENV.LOCAL]: 123,
    },
    BERACHAIN: {
        [constants_1.ENV.PROD]: 0,
        [constants_1.ENV.STAGING]: 80085,
        [constants_1.ENV.DEV]: 80085,
        [constants_1.ENV.LOCAL]: 80085,
    },
    LINEA: {
        [constants_1.ENV.PROD]: 59144,
        [constants_1.ENV.STAGING]: 59141,
        [constants_1.ENV.DEV]: 59141,
        [constants_1.ENV.LOCAL]: 59141,
    },
    CYBERCONNECT: {
        [constants_1.ENV.PROD]: 7560,
        [constants_1.ENV.STAGING]: 111557560,
        [constants_1.ENV.DEV]: 111557560,
        [constants_1.ENV.LOCAL]: 111557560,
    },
    BASE: {
        [constants_1.ENV.PROD]: 8453,
        [constants_1.ENV.STAGING]: 84532,
        [constants_1.ENV.DEV]: 84532,
        [constants_1.ENV.LOCAL]: 84532,
    },
};
exports.CHAIN_ID = Object.assign({ ETHEREUM: exports.ETH_CHAIN_ID }, exports.ALIAS_CHAIN_ID);
exports.CHAIN_NAME = {
    // eth
    1: 'ETHEREUM',
    11155111: 'ETHEREUM',
    // polygon
    137: 'POLYGON',
    80002: 'POLYGON',
    // bsc
    56: 'BSC',
    97: 'BSC',
    // optimism
    10: 'OPTIMISM',
    11155420: 'OPTIMISM',
    // plygonzkevm
    1101: 'POLYGONZKEVM',
    2442: 'POLYGONZKEVM',
    // arbitrun
    421614: 'ARBITRUN',
    42161: 'ARBITRUM',
    // fuse
    122: 'FUSE',
    123: 'FUSE',
    // berachain
    80085: 'BERACHAIN',
    // linea
    59144: 'LINEA',
    59141: 'LINEA',
    // cyberconnect
    7560: 'CYBER_CONNECT_MAINNET',
    111557560: 'CYBER_CONNECT_TESTNET',
    // base
    8453: 'BASE_MAINNET',
    84532: 'BASE_TESTNET',
};
exports.VIEM_CORE_CONFIG = {
    [constants_1.ENV.PROD]: {
        NETWORK: chains_1.mainnet,
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
        EPNS_CORE_CONTRACT: '0x66329Fdd4042928BfCAB60b179e1538D56eeeeeE',
    },
    [constants_1.ENV.STAGING]: {
        NETWORK: chains_1.sepolia,
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
        EPNS_CORE_CONTRACT: '0x9d65129223451fbd58fc299c635cd919baf2564c',
    },
    [constants_1.ENV.DEV]: {
        NETWORK: chains_1.sepolia,
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
        EPNS_CORE_CONTRACT: '0x5ab1520e2bd519bdab2e1347eee81c00a77f4946',
    },
    [constants_1.ENV.LOCAL]: {
        NETWORK: chains_1.sepolia,
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
        EPNS_CORE_CONTRACT: '0x5ab1520e2bd519bdab2e1347eee81c00a77f4946',
    },
};
exports.CORE_CONFIG = {
    [constants_1.ENV.PROD]: {
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
        EPNS_CORE_CONTRACT: '0x66329Fdd4042928BfCAB60b179e1538D56eeeeeE',
    },
    [constants_1.ENV.STAGING]: {
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
        EPNS_CORE_CONTRACT: '0x9d65129223451fbd58fc299c635cd919baf2564c',
    },
    [constants_1.ENV.DEV]: {
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
        EPNS_CORE_CONTRACT: '0x5ab1520e2bd519bdab2e1347eee81c00a77f4946',
    },
    [constants_1.ENV.LOCAL]: {
        API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
        EPNS_CORE_CONTRACT: '0x5ab1520e2bd519bdab2e1347eee81c00a77f4946',
    },
};
const CONFIG = {
    [constants_1.ENV.PROD]: {
        [BLOCKCHAIN_NETWORK.ETH_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BSC_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUMONE_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.FUSE_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.LINEA_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0x0d8e75CB5d8873c43c5d9Add71Fd71a09F7Ef890',
        },
        [BLOCKCHAIN_NETWORK.BASE_MAINNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        }
    },
    [constants_1.ENV.STAGING]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x0c34d54a09cfe75bccd878a469206ae77e0fe6e7',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9Dc25996ba72A2FD7E64e7a674232a683f406F1A',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489b7af21ceb969f49a90e481274966ce9d74d',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9Dc25996ba72A2FD7E64e7a674232a683f406F1A',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x7b9C405e261ba671f008c20D0321f62d08C140EC',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489B7af21cEb969f49A90E481274966ce9D74d',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489B7af21cEb969f49A90E481274966ce9D74d',
        }
    },
    [constants_1.ENV.DEV]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9dDCD7ed7151afab43044E4D694FA064742C428c',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x4132061E3349ff36cFfCadA460E10Bd4f31F7ea8',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x7eBb54D86CF928115965DB596a3E600404dD8039',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0xA1DF3E68D085aa6918bcc2506b24e499830Db0eB',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        }
    },
    [constants_1.ENV.LOCAL]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9dDCD7ed7151afab43044E4D694FA064742C428c',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x4132061E3349ff36cFfCadA460E10Bd4f31F7ea8',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x7eBb54D86CF928115965DB596a3E600404dD8039',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0xA1DF3E68D085aa6918bcc2506b24e499830Db0eB',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        }
    },
};
exports.default = CONFIG;
exports.TOKEN = {
    [constants_1.ENV.PROD]: '0xf418588522d5dd018b425E472991E52EBBeEEEEE',
    [constants_1.ENV.STAGING]: '0x37c779a1564DCc0e3914aB130e0e787d93e21804',
    [constants_1.ENV.DEV]: '0x37c779a1564DCc0e3914aB130e0e787d93e21804',
    [constants_1.ENV.LOCAL]: '0x37c779a1564DCc0e3914aB130e0e787d93e21804',
};
exports.TOKEN_VIEM_NETWORK_MAP = {
    [constants_1.ENV.PROD]: chains_1.mainnet,
    [constants_1.ENV.STAGING]: chains_1.sepolia,
    [constants_1.ENV.DEV]: chains_1.sepolia,
    [constants_1.ENV.LOCAL]: chains_1.sepolia,
};
exports.MIN_TOKEN_BALANCE = {
    [constants_1.ENV.PROD]: 50,
    [constants_1.ENV.STAGING]: 50,
    [constants_1.ENV.DEV]: 50,
    [constants_1.ENV.LOCAL]: 50,
};
exports.ABIS = {
    CORE: core_1.coreABI,
    COMM: comm_1.commABI,
    TOKEN: token_1.tokenABI,
};
exports.CHANNEL_TYPE = {
    TIMEBOUND: 4,
    GENERAL: 2,
};
exports.VIEM_CONFIG = {
    [constants_1.ENV.PROD]: {
        [BLOCKCHAIN_NETWORK.ETH_MAINNET]: {
            NETWORK: chains_1.mainnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_MAINNET]: {
            NETWORK: chains_1.polygon,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BSC_MAINNET]: {
            NETWORK: chains_1.bsc,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_MAINNET]: {
            NETWORK: chains_1.optimism,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_MAINNET]: {
            NETWORK: chains_1.polygonZkEvm,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUMONE_MAINNET]: {
            NETWORK: chains_1.arbitrum,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.FUSE_MAINNET]: {
            NETWORK: chains_1.fuse,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.LINEA_MAINNET]: {
            NETWORK: chains_1.linea,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0x0d8e75CB5d8873c43c5d9Add71Fd71a09F7Ef890',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_MAINNET]: {
            NETWORK: customChains_1.cyberConnectMainnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BASE_MAINNET]: {
            NETWORK: chains_1.base,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.PROD],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        }
    },
    [constants_1.ENV.STAGING]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            NETWORK: chains_1.sepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x0c34d54a09cfe75bccd878a469206ae77e0fe6e7',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            NETWORK: customChains_1.polygonAmoy,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            NETWORK: chains_1.bscTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            NETWORK: chains_1.optimismSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9Dc25996ba72A2FD7E64e7a674232a683f406F1A',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            NETWORK: customChains_1.polygonZkEvmCordona,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489b7af21ceb969f49a90e481274966ce9d74d',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            NETWORK: chains_1.arbitrumSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9Dc25996ba72A2FD7E64e7a674232a683f406F1A',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            NETWORK: chains_1.fuseSparknet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0xb3971BCef2D791bc4027BbfedFb47319A4AAaaAa',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            NETWORK: customChains_1.berachainTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x7b9C405e261ba671f008c20D0321f62d08C140EC',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            NETWORK: customChains_1.lineaSepoliaTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            NETWORK: customChains_1.cyberConnectTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489B7af21cEb969f49A90E481274966ce9D74d',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            NETWORK: chains_1.baseSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.STAGING],
            EPNS_COMMUNICATOR_CONTRACT: '0x6e489B7af21cEb969f49A90E481274966ce9D74d',
        }
    },
    [constants_1.ENV.DEV]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            NETWORK: chains_1.sepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9dDCD7ed7151afab43044E4D694FA064742C428c',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            NETWORK: customChains_1.polygonAmoy,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            NETWORK: chains_1.bscTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x4132061E3349ff36cFfCadA460E10Bd4f31F7ea8',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            NETWORK: chains_1.optimismSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            NETWORK: customChains_1.polygonZkEvmCordona,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            NETWORK: chains_1.arbitrumSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            NETWORK: chains_1.fuseSparknet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x7eBb54D86CF928115965DB596a3E600404dD8039',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            NETWORK: customChains_1.berachainTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0xA1DF3E68D085aa6918bcc2506b24e499830Db0eB',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            NETWORK: customChains_1.lineaSepoliaTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            NETWORK: customChains_1.cyberConnectTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            NETWORK: chains_1.baseSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.DEV],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        }
    },
    [constants_1.ENV.LOCAL]: {
        [BLOCKCHAIN_NETWORK.ETH_SEPOLIA]: {
            NETWORK: chains_1.sepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9dDCD7ed7151afab43044E4D694FA064742C428c',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_AMOY]: {
            NETWORK: customChains_1.polygonAmoy,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.BSC_TESTNET]: {
            NETWORK: chains_1.bscTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x4132061E3349ff36cFfCadA460E10Bd4f31F7ea8',
        },
        [BLOCKCHAIN_NETWORK.OPTIMISM_TESTNET]: {
            NETWORK: chains_1.optimismSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.POLYGON_ZK_EVM_TESTNET]: {
            NETWORK: customChains_1.polygonZkEvmCordona,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550b5c92baa056fc0f08132f49508145f',
        },
        [BLOCKCHAIN_NETWORK.ARBITRUM_TESTNET]: {
            NETWORK: chains_1.arbitrumSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x754787358fac861ef904c92d54f7adb659779317',
        },
        [BLOCKCHAIN_NETWORK.FUSE_TESTNET]: {
            NETWORK: chains_1.fuseSparknet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x7eBb54D86CF928115965DB596a3E600404dD8039',
        },
        [BLOCKCHAIN_NETWORK.BERACHAIN_TESTNET]: {
            NETWORK: customChains_1.berachainTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0xA1DF3E68D085aa6918bcc2506b24e499830Db0eB',
        },
        [BLOCKCHAIN_NETWORK.LINEA_TESTNET]: {
            NETWORK: customChains_1.lineaSepoliaTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.CYBER_CONNECT_TESTNET]: {
            NETWORK: customChains_1.cyberConnectTestnet,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        },
        [BLOCKCHAIN_NETWORK.BASE_TESTNET]: {
            NETWORK: chains_1.baseSepolia,
            API_BASE_URL: exports.API_BASE_URL[constants_1.ENV.LOCAL],
            EPNS_COMMUNICATOR_CONTRACT: '0x9cb3bd7550B5c92baA056Fc0F08132f49508145F',
        }
    },
};
exports.ALPHA_FEATURE_CONFIG = {
    STABLE: {
        feature: [],
    },
    ALPHA: {
        feature: [constants_1.default.ALPHA_FEATURES.SCALABILITY_V2],
    },
};
